IndexOutOfBoundsException in ORDER BY comparator during temporal range queries

## Summary

An `IndexOutOfBoundsException` occurs in the ORDER BY operator when comparing rows across pages during a temporal range query with `ORDER BY ... DESC`.

The error manifests as an out-of-bounds Arrow buffer access:

```
ERROR: index: 348784, length: 8 (expected: range(0, 22232))
Detail: {"@type":"xt:fault","@value":{"xtdb.error/message":"index: 348784, length: 8 (expected: range(0, 22232))","xtdb.error/data":{"xtdb.error/code":{"@type":"xt:keyword","@value":"xtdb.error/unknown"}}}}
```

> **Note:** This does **not** fail when rows have distinct IDs — the issue requires multiple versions against a single ID.

## Query Pattern

```sql
FROM <table>
    FOR VALID_TIME
        FROM TIMESTAMP '2026-02-20T02:30:00Z'
        TO TIMESTAMP '2026-02-20T08:30:00Z'
SELECT _id, _valid_time, _valid_from
ORDER BY _valid_from DESC
```

<details>
<summary>Original Stack Trace</summary>

```
xtdb.error.Fault: index: 348784, length: 8 (expected: range(0, 22232))
    at xtdb.error$fault.invokeStatic(error.clj:70)
    ...
Caused by: java.lang.IndexOutOfBoundsException: index: 348784, length: 8 (expected: range(0, 22232))
    at org.apache.arrow.memory.ArrowBuf.checkIndexD(ArrowBuf.java:299)
    at org.apache.arrow.memory.ArrowBuf.getLong(ArrowBuf.java:312)
    at xtdb.arrow.ExtensibleBuffer.getLong(Buffers.kt:92)
    at xtdb.arrow.FixedWidthVector.getLong(FixedWidthVector.kt:132)
    at xtdb.arrow.TimestampTzVector.getLong(TimestampVector.kt:81)
    at xtdb.arrow.ValueReader$ForVector.readLong(ValueReader.kt:37)
    at clojure.core$eval39888$fn$reify__38992.applyAsInt(NO_SOURCE_FILE:0)
    at xtdb.operator.order_by$mk_rel_comparator$fn__18486$G__18491__18492.invoke(order_by.clj:89)
    at clojure.lang.AFunction.compare(AFunction.java:53)
```

</details>

<details>
<summary>Local Replication — Integration Test</summary>

The issue has been replicated end-to-end with the following integration test. The external sort / k-way merge path in the order-by operator is triggered because the default `*block-size*` is 102400, requiring 330k rows.

```clojure
(t/deftest ^:integration test-order-by-temporal-range-desc
  (with-open [node (xtn/start-node (merge tu/*node-opts*
                                          {:compactor {:threads 0}}))]
    (let [base-vf (Instant/parse "2020-01-01T00:00:00Z")]
      ;; 330k rows, each with a distinct valid-from 1 minute apart
      (doseq [batch (partition-all 1000 (range 330000))]
        (xt/execute-tx node (mapv (fn [i]
                                    [:put-docs {:into :docs
                                                :valid-from (.plus base-vf (Duration/ofMinutes i))}
                                     {:xt/id "doc1" :val i}])
                                  batch))))

    (let [results (xt/q node
                        (str "FROM docs"
                             "  FOR VALID_TIME"
                             "    FROM TIMESTAMP '2020-01-01T00:00:00Z'"
                             "    TO TIMESTAMP '2021-01-01T00:00:00Z'"
                             " SELECT _id, _valid_from"
                             " ORDER BY _valid_from DESC"))]
      (t/is (= 330000 (count results)))
      (t/is (= #xt/zdt "2020-03-17T08:23Z[UTC]"
               (:xt/valid-from (first results))))
      (t/is (= #xt/zdt "2020-01-01T00:00Z[UTC]"
               (:xt/valid-from (last results)))))))
```

</details>

<details>
<summary>Stack Trace from Integration Test</summary>

```
15:21:32.630 [pgwire-connection--pool-17-thread-37] ERROR xtdb.pgwire - Uncaught exception processing message
xtdb.error.Fault: index: 168000, length: 8 (expected: range(0, 168000))
    at xtdb.error$fault.invokeStatic(error.clj:70)
    at xtdb.error$fault.invoke(error.clj:62)
    at xtdb.error$fn__2951.invokeStatic(error.clj:184)
    at xtdb.error$fn__2951.invoke(error.clj:153)
    at xtdb.error$fn__2896$G__2891__2903.invoke(error.clj:14)
    at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:276)
    at xtdb.pgwire$send_ex.invoke(pgwire.clj:270)
    at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1733)
    at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1711)
    at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1764)
    at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1735)
    at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1810)
    at xtdb.pgwire$connect.invoke(pgwire.clj:1768)
    at xtdb.pgwire$accept_loop$fn__14549$fn__14550$fn__14551.invoke(pgwire.clj:1857)
    at clojure.lang.AFn.run(AFn.java:22)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
    at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
    at java.base/java.lang.Thread.run(Unknown Source)
```

</details>

--- a/core/src/main/clojure/xtdb/operator/order_by.clj
+++ b/core/src/main/clojure/xtdb/operator/order_by.clj
@@ -85,9 +85,11 @@
                   col-comparator (expr.comp/->comparator read-col1 read-col2 null-ordering)
 
                   ^Comparator
-                  comparator (cond-> ^Comparator (fn [left right]
-                                                   (.applyAsInt col-comparator left right))
-                               (= :desc direction) (.reversed))]
+                  comparator (if (= :desc direction)
+                               (fn [left right]
+                                 (- (.applyAsInt col-comparator left right)))
+                               (fn [left right]
+                                 (.applyAsInt col-comparator left right)))]
               (if acc
                 (.thenComparing acc comparator)
                 comparator)))

--- a/src/test/clojure/xtdb/operator/order_by_test.clj
+++ b/src/test/clojure/xtdb/operator/order_by_test.clj
@@ -4,7 +4,7 @@
             [xtdb.node :as xtn]
             [xtdb.operator.order-by :as order-by]
             [xtdb.test-util :as tu])
-  (:import (java.time Instant Duration)))
+  (:import (java.time Instant Duration ZoneId)))
 
 (t/use-fixtures :each tu/with-allocator)
 
@@ -85,20 +85,15 @@
                                       [:put-docs {:into :docs
                                                   :valid-from (.plus base-vf (Duration/ofMinutes i))}
                                        {:xt/id "doc1" :val i}])
-                                    batch))))
+                                    batch)))
 
-      (t/testing "ORDER BY valid from ascending"
-        (let [results-asc (xt/q node "FROM docs FOR VALID_TIME ALL SELECT id, _valid_from ORDER BY _valid_from")]
-          (t/is (= 24 (count results-asc)))
-          (t/is (= #xt/zdt "2020-01-01T00:00Z[UTC]"
-                   (:xt/valid-from (first results-asc))))
-          (t/is (= #xt/zdt "2020-01-01T00:23Z[UTC]"
-                   (:xt/valid-from (last results-asc))))))
+        (let [expected-vfs (mapv #(.atZone (.plus base-vf (Duration/ofMinutes %)) (ZoneId/of "UTC")) (range 24))]
+          (t/testing "ORDER BY valid from ascending"
+            (let [results (xt/q node "FROM docs FOR VALID_TIME ALL SELECT _id, _valid_from ORDER BY _valid_from")]
+              (t/is (= expected-vfs
+                       (mapv :xt/valid-from results)))))
 
-      (t/testing "ORDER BY valid from descending"
-        (let [results-desc (xt/q node "FROM docs FOR VALID_TIME ALL SELECT id, _valid_from ORDER BY _valid_from DESC")]
-          (t/is (= 24 (count results-desc)))
-          (t/is (= #xt/zdt "2020-01-01T00:23Z[UTC]"
-                   (:xt/valid-from (first results-desc))))
-          (t/is (= #xt/zdt "2020-01-01T00:00Z[UTC]"
-                   (:xt/valid-from (last results-desc)))))))))
+          (t/testing "ORDER BY valid from descending"
+            (let [results (xt/q node "FROM docs FOR VALID_TIME ALL SELECT _id, _valid_from ORDER BY _valid_from DESC")]
+              (t/is (= (rseq expected-vfs)
+                       (mapv :xt/valid-from results))))))))))

--- a/src/test/clojure/xtdb/operator/order_by_test.clj
+++ b/src/test/clojure/xtdb/operator/order_by_test.clj
@@ -1,7 +1,10 @@
 (ns xtdb.operator.order-by-test
   (:require [clojure.test :as t]
+            [xtdb.api :as xt]
+            [xtdb.node :as xtn]
             [xtdb.operator.order-by :as order-by]
-            [xtdb.test-util :as tu]))
+            [xtdb.test-util :as tu])
+  (:import (java.time Instant Duration)))
 
 (t/use-fixtures :each tu/with-allocator)
 
@@ -41,19 +44,19 @@
   (let [table-with-nil [{:a 12.4, :b 10}, {:a nil, :b 15}, {:a 100, :b 83}, {:a 83.0, :b 100}]]
     (t/is (= [{:b 15}, {:a 12.4, :b 10}, {:a 83.0, :b 100}, {:a 100, :b 83}]
              (tu/query-ra '[:order-by {:order-specs [[a {:null-ordering :nulls-first}]]}
-                          [:table {:param ?table}]]
+                            [:table {:param ?table}]]
                           {:args {:table table-with-nil}}))
           "nulls first")
 
     (t/is (= [{:a 12.4, :b 10}, {:a 83.0, :b 100}, {:a 100, :b 83}, {:b 15}]
              (tu/query-ra '[:order-by {:order-specs [[a {:null-ordering :nulls-last}]]}
-                          [:table {:param ?table}]]
+                            [:table {:param ?table}]]
                           {:args {:table table-with-nil}}))
           "nulls last")
 
     (t/is (= [{:a 12.4, :b 10}, {:a 83.0, :b 100}, {:a 100, :b 83}, {:b 15}]
              (tu/query-ra '[:order-by {:order-specs [[a]]}
-                          [:table {:param ?table}]]
+                            [:table {:param ?table}]]
                           {:args {:table table-with-nil}}))
           "default nulls last")))
 
@@ -67,3 +70,35 @@
                              [::tu/pages batches]]
                             {}))
             "spilling to disk"))))
+
+(t/deftest test-order-by-temporal-range-5269
+  ;; #5269 — IOOBE in ORDER BY DESC with external sort.
+  ;; .reversed() on the k-way merge comparator swapped index arguments between
+  ;; relations with different row counts, reading past the smaller one's buffer.
+  (with-redefs [order-by/*block-size* (int 10)]
+    (with-open [node (xtn/start-node (merge tu/*node-opts*
+                                            {:compactor {:threads 0}
+                                             :indexer {:rows-per-block 10}}))]
+      (let [base-vf (Instant/parse "2020-01-01T00:00:00Z")]
+        (doseq [batch (partition-all 10 (range 24))]
+          (xt/execute-tx node (mapv (fn [i]
+                                      [:put-docs {:into :docs
+                                                  :valid-from (.plus base-vf (Duration/ofMinutes i))}
+                                       {:xt/id "doc1" :val i}])
+                                    batch))))
+
+      (t/testing "ORDER BY valid from ascending"
+        (let [results-asc (xt/q node "FROM docs FOR VALID_TIME ALL SELECT id, _valid_from ORDER BY _valid_from")]
+          (t/is (= 24 (count results-asc)))
+          (t/is (= #xt/zdt "2020-01-01T00:00Z[UTC]"
+                   (:xt/valid-from (first results-asc))))
+          (t/is (= #xt/zdt "2020-01-01T00:23Z[UTC]"
+                   (:xt/valid-from (last results-asc))))))
+
+      (t/testing "ORDER BY valid from descending"
+        (let [results-desc (xt/q node "FROM docs FOR VALID_TIME ALL SELECT id, _valid_from ORDER BY _valid_from DESC")]
+          (t/is (= 24 (count results-desc)))
+          (t/is (= #xt/zdt "2020-01-01T00:23Z[UTC]"
+                   (:xt/valid-from (first results-desc))))
+          (t/is (= #xt/zdt "2020-01-01T00:00Z[UTC]"
+                   (:xt/valid-from (last results-desc)))))))))

